spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDelete.scala

```diff
@@ -0,0 +1,152 @@
+import org.apache.spark.sql.execution.datasources.DataSourceStrategy
+import org.apache.spark.sql.execution.datasources.v2.{DataSourceV2Relation, DataSourceV2ScanRelation, PushDownUtils}
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types.{BooleanType, StructType}
+import org.apache.spark.sql.util.CaseInsensitiveStringMap
+
+// TODO: should be part of early scan push down after the delete condition is optimized
+object RewriteDelete extends Rule[LogicalPlan] with PredicateHelper with Logging {
+
+  import org.apache.spark.sql.execution.datasources.v2.ExtendedDataSourceV2Implicits._
+
+  private val FILE_NAME_COL = "_file"
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
+    // don't rewrite deletes that can be answered by passing filters to deleteWhere in SupportsDelete
+    case d @ DeleteFromTable(r: DataSourceV2Relation, Some(cond)) if isDeleteWhereCase(r, cond) =>
+      d
+
+    // rewrite all operations that require reading the table to delete records
+    case DeleteFromTable(r: DataSourceV2Relation, Some(cond)) =>
+      // TODO: do a switch based on whether we get BatchWrite or DeltaBatchWrite
+      val writeInfo = newWriteInfo(r.schema)
+      val mergeBuilder = r.table.asMergeable.newMergeBuilder(writeInfo)
+
+      val scanPlan = buildScanPlan(r.table, r.output, mergeBuilder, cond)
+
+      val remainingRowFilter = Not(EqualNullSafe(cond, Literal(true, BooleanType)))
+      val remainingRowsPlan = Filter(remainingRowFilter, scanPlan)
+
+      val batchWrite = mergeBuilder.asWriteBuilder.buildForBatch()
+      val writePlan = buildWritePlan(remainingRowsPlan, r.output)
+      ReplaceData(r, batchWrite, writePlan)
+  }
+
+  private def buildScanPlan(
+      table: Table,
+      output: Seq[AttributeReference],
+      mergeBuilder: MergeBuilder,
+      cond: Expression): LogicalPlan = {
+
+    val scanBuilder = mergeBuilder.asScanBuilder
+
+    val predicates = splitConjunctivePredicates(cond)
+    val normalizedPredicates = DataSourceStrategy.normalizeExprs(predicates, output)
+    PushDownUtils.pushFilters(scanBuilder, normalizedPredicates)
+
+    val scan = scanBuilder.build()
+    val scanRelation = DataSourceV2ScanRelation(table, scan, output)
+
+    val scanPlan = scan match {
+      case _: SupportsFileFilter =>
+        val matchingFilePlan = buildFileFilterPlan(cond, scanRelation)
+        val dynamicFileFilter = DynamicFileFilter(scanRelation, matchingFilePlan)
+        dynamicFileFilter
+      case _ =>
+        scanRelation
+    }
+
+    // include file name so that we can group data back
+    val fileNameExpr = Alias(InputFileName(), FILE_NAME_COL)()
+    Project(scanPlan.output :+ fileNameExpr, scanPlan)
+  }
+
+  private def buildWritePlan(
+      remainingRowsPlan: LogicalPlan,
+      output: Seq[AttributeReference]): LogicalPlan = {
+
+    // TODO: sort by _pos to keep the original ordering of rows
+    // TODO: consider setting a file size limit
```

Comment:
I want to bring attention to these TODOs. It would require a bit of effort for the first one but the second one is something we can do now. The main question is whether we want to allow any extra size overhead before closing files. For example, there may be a file with 1.1 GB of data and our target size can be 1GB. Having a hard limit would mean we will cut 1.1 GB file into two.

Reply:
The problem with such limits is that we never know what would be better. There may be an opposite situation when we had 1 GB file and added extra 100 MB from another file and wrote together. If we want to really solve it, we have to adapt our writers so that we can close the current file if `_file` column changes.

Reply:
The problem with a soft limit is that you have a hard limit somewhere. You could have 1.2 GB of data and the soft limit would make you cut at 1.1 GB. I think the issue you can't avoid is not knowing how much more data there is coming from the writer. I would not spend time worrying about more complicated logic here. We can still tune the files after the fact with rewrites.

This problem also has less of an impact if we get the _pos changes done, so I would focus effort there. Keeping the original order will help keep files the same size, and we can always group just the rows from one input file into an output file.